channelserver/src/session.rs

```diff
@@ -191,3 +207,40 @@ impl StreamHandler<ws::Message, ws::ProtocolError> for WsChannelSession {
         }
     }
 }
+
+impl WsChannelSession {
+    fn hb(&self, ctx: &mut ws::WebsocketContext<Self, WsChannelSessionState>) {
+        ctx.run_interval(HEARTBEAT_INTERVAL, |act, ctx| {
+            // Has the connection's lifespan expired?
+            if Instant::now() > act.expiry {
+                ctx.state().log.do_send(logging::LogMessage {
+                    level: logging::ErrorLevel::Warn,
+                    msg: format!("Client connected too long. {}:{}", act.id, act.channel,),
+                });
+                ctx.state().addr.do_send(server::Disconnect {
+                    id: act.id,
+                    channel: act.channel.clone(),
+                    reason: server::DisconnectReason::Timeout,
+                });
+                ctx.stop();
+                return;
+            }
+            // Have we not gotten any traffic from the client?
+            if Instant::now().duration_since(act.hb) > CLIENT_TIMEOUT {
```

Comment:
It looks like `act.hb` is updated on ` ws::Message::{Ping, Pong}` as well as `ws::Message::Text`.  Is that the desired behaviour?  Naively, I would expect the ping/pong keepalives not to count as "activity" for the purposes of this client-activity timeout, since they don't really correspond to the client "doing" anything and might be automatically sent by the websocket framework.

Reply:
*nod* make sense, thanks.